core/src/main/java/org/jruby/ir/IRBuilder.java

```diff
@@ -3980,75 +3982,77 @@ public Operand buildZArray(Variable result) {
+        IRScope superScope = scope;
+
+        boolean defineMethod = false;
+        // Figure out depth from argument scope and whether defineMethod may be one of the method calls.
+        while (superScope instanceof IRClosure) {
+            if (superBuilder != null && superBuilder.isMethodDefine()) defineMethod = true;
+
+            // We may run out of live builds and walk int already built scopes if zsuper in an eval
+            superBuilder = superBuilder != null && superBuilder.parent != null ? superBuilder.parent : null;
+            superScope = superScope.getLexicalParent();
+            depthFromSuper++;
+        }
+
+        // If we hit a method, this is known to always succeed
+        Variable zsuperResult = createTemporaryVariable();
+        if (superScope instanceof IRMethod && !defineMethod) {
+            Operand[] args = adjustVariableDepth(getCallArgs(superScope, superBuilder), depthFromSuper);
+            addInstr(new ZSuperInstr(scope, zsuperResult, buildSelf(), args, block, scope.maybeUsingRefinements()));
+        } else {
+            // We will not have a zsuper show up since we won't emit it but we still need to toggle it.
+            // define_method optimization will try and create a method from a closure but it should not in this case.
+            scope.setUsesZSuper();
 
-                addInstr(new LabelInstr(next));
+            // Two conditions will inject an error:
+            // 1. We cannot find any method scope above the closure (e.g. module A; define_method(:a) { super }; end)
+            // 2. One of the method calls the closure is passed to is named define_method.
+            //
+            // Note: We are introducing an issue but it is so obscure we are ok with it.
+            // A method named define_method containing zsuper in a method scope which is not actually
+            // a define_method will get raised as invalid even though it should zsuper to the method.
+            addRaiseError("RuntimeError",
+                    "implicit argument passing of super from method defined by define_method() is not supported. Specify all arguments explicitly.");
+        }
 
-                // If we hit a method, this is known to always succeed
-                if (superScope instanceof IRMethod) {
-                    Operand[] args = adjustVariableDepth(getCallArgs(superScope, superBuilder), depthFromSuper);
-                    addInstr(new ZSuperInstr(scope, zsuperResult, buildSelf(), args, block, scope.maybeUsingRefinements()));
-                } //else {
-                // FIXME: Do or don't ... there is no try
-                    /* Control should never get here in the runtime */
-                    /* Should we add an exception throw here just in case? */
-                //}
+        return zsuperResult;
+    }
 
-                addInstr(new LabelInstr(allDoneLabel));
-                return zsuperResult;
-            }
-        };
+    private boolean isMethodDefine() {
```

Comment:
Not `isDefineMethod`?

Reply:
yeah I yoda named that for no reason.   I will change that.